File: day20_part1_2022.java

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

class Solution {
    static class Num {
        int pos;
        int val;

        Num(int pos, int val) {
            this.pos = pos;
            this.val = val;
        }
    }

    public static void main(String[] args) {
        List<Num> nums = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader("input.txt"))) {
            String line;
            int i = 0;
            while ((line = br.readLine()) != null) {
                nums.add(new Num(i, Integer.parseInt(line)));
                i++;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        List<Num> nums2 = new ArrayList<>(nums.size());
        for (Num num : nums) {
            nums2.add(new Num(num.pos, 811589153 * num.val));
        }

        mix(nums);

        System.out.println(coords(nums));
    }

    static void mix(List<Num> nums) {
        int n = nums.size() - 1;
        for (Num num : nums) {
            int oldPos = num.pos;
            int newPos = ((oldPos + num.val) % n + n) % n;
            if (oldPos < newPos) {
                for (Num num2 : nums) {
                    if (num2.pos > oldPos && num2.pos <= newPos) {
                        num2.pos--;
                    }
                }
            }
            if (newPos < oldPos) {
                for (Num num2 : nums) {
                    if (num2.pos >= newPos && num2.pos < oldPos) {
                        num2.pos++;
                    }
                }
            }
            num.pos = newPos;
        }
    }

    static int coords(List<Num> nums) {
        int l = nums.size();
        int zeroPos = 0;
        for (Num num : nums) {
            if (num.val == 0) {
                zeroPos = num.pos;
                break;
            }
        }
        int sum = 0;
        for (Num num : nums) {
            if (num.pos == (zeroPos + 1000) % l || num.pos == (zeroPos + 2000) % l || num.pos == (zeroPos + 3000) % l) {
                sum += num.val;
            }
        }
        return sum;
    }
}

File: objectivec/day20_part1_2022.m

#import <Foundation/Foundation.h>

typedef struct {
    long long pos;
    long long val;
} Num;

static void mix(Num *nums, NSUInteger size) {
    long long n = size - 1;
    for (NSUInteger i = 0; i < size; ++i) {
        long long oldPos = nums[i].pos;
        long long newPos = ((oldPos + nums[i].val) % n + n) % n;
        if (oldPos < newPos) {
            for (NSUInteger j = 0; j < size; ++j) {
                if (nums[j].pos > oldPos && nums[j].pos <= newPos) {
                    nums[j].pos--;
                }
            }
        } else if (newPos < oldPos) {
            for (NSUInteger j = 0; j < size; ++j) {
                if (nums[j].pos >= newPos && nums[j].pos < oldPos) {
                    nums[j].pos++;
                }
            }
        }
        nums[i].pos = newPos;
    }
}

static long long coords(Num *nums, NSUInteger size) {
    long long zeroPos = 0;
    for (NSUInteger i = 0; i < size; ++i) {
        if (nums[i].val == 0) {
            zeroPos = nums[i].pos;
            break;
        }
    }
    long long sum = 0;
    for (NSUInteger i = 0; i < size; ++i) {
        long long p = nums[i].pos;
        if (p == (zeroPos + 1000) % size ||
            p == (zeroPos + 2000) % size ||
            p == (zeroPos + 3000) % size) {
            sum += nums[i].val;
        }
    }
    return sum;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *path = @"input.txt";
        NSError *err = nil;
        NSString *content = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&err];
        if (!content) {
            fprintf(stderr, "Error opening file: %s\n", err.localizedDescription.UTF8String);
            return -1;
        }

        NSArray<NSString *> *lines = [content componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]];
        NSUInteger size = 0;
        Num *nums = NULL;

        for (NSString *line in lines) {
            if (line.length == 0) continue;
            long long v = line.longLongValue;
            Num num = { .pos = size, .val = v };
            ++size;
            nums = realloc(nums, size * sizeof(Num));
            nums[size - 1] = num;
        }

        mix(nums, size);
        long long result = coords(nums, size);
        printf("%lld\n", result);

        free(nums);
    }
    return 0;
}
